modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/PlannerTest.java

```diff
@@ -2656,6 +2659,221 @@ public void testJoinPushExpressionRule() throws Exception {
+            RewindabilityTraitDef.INSTANCE,
+            CorrelationTraitDef.INSTANCE
+        };
+
+        PlanningContext ctx = PlanningContext.builder()
+            .localNodeId(F.first(nodes))
+            .originatingNodeId(F.first(nodes))
+            .parentContext(Contexts.empty())
+            .frameworkConfig(newConfigBuilder(FRAMEWORK_CONFIG)
+                .defaultSchema(schema)
+                .traitDefs(traitDefs)
+                .build())
+            .logger(log)
+            .query(sql)
+            .topologyVersion(AffinityTopologyVersion.NONE)
+            .build();
+
+        try (IgnitePlanner planner = ctx.planner()) {
+            assertNotNull(planner);
+
+            String qry = ctx.query();
+
+            assertNotNull(qry);
+
+            // Parse
+            SqlNode sqlNode = planner.parse(qry);
+
+            // Validate
+            sqlNode = planner.validate(sqlNode);
+
+            // Convert to Relational operators graph
+            RelRoot relRoot = planner.rel(sqlNode);
+
+            RelNode rel = relRoot.rel;
+
+            assertNotNull(rel);
+            assertEquals("" +
+                    "LogicalSort(sort0=[$3], sort1=[$0], dir0=[ASC], dir1=[ASC])\n" +
+                    "  LogicalProject(DEPTNO=[$0], NAME=[$1], ID=[$2], NAME0=[$3], DEPTNO0=[$4])\n" +
+                    "    LogicalJoin(condition=[AND(=($0, $4), =($3, $1))], joinType=[inner])\n" +
+                    "      IgniteLogicalTableScan(table=[[PUBLIC, DEPT]])\n" +
+                    "      IgniteLogicalTableScan(table=[[PUBLIC, EMP]])\n",
+                RelOptUtil.toString(rel));
+
+            // Transformation chain
+            RelTraitSet desired = rel.getTraitSet()
+                .replace(IgniteConvention.INSTANCE)
+                .replace(IgniteDistributions.single())
+                .simplify();
+
+            RelNode phys = planner.transform(PlannerPhase.OPTIMIZATION, desired, rel);
+
+            assertNotNull(phys);
+            assertEquals("" +
+                    "IgniteMergeJoin(condition=[AND(=($0, $4), =($3, $1))], joinType=[inner])\n" +
+                    "  IgniteIndexScan(table=[[PUBLIC, DEPT]], index=[dep_idx])\n" +
+                    "  IgniteIndexScan(table=[[PUBLIC, EMP]], index=[emp_idx])\n",
+                RelOptUtil.toString(phys));
+        }
+    }
+
+    /** */
+    @Test
+    public void testMergeJoinIsNotAppliedForNonEquiJoin() throws Exception {
+        IgniteTypeFactory f = new IgniteTypeFactory(IgniteTypeSystem.INSTANCE);
+
+        TestTable emp = new TestTable(
+            new RelDataTypeFactory.Builder(f)
+                .add("ID", f.createJavaType(Integer.class))
+                .add("NAME", f.createJavaType(String.class))
+                .add("DEPTNO", f.createJavaType(Integer.class))
+                .build()) {
+
+            @Override public IgniteDistribution distribution() {
+                return IgniteDistributions.broadcast();
```

Comment:
I found, that changing distribution into : return IgniteDistributions.affinity(0, "EMP", "hash");
will not return physical plan at all, is it ok ?

Reply:
fixed

Reply:
ok, you are right.